### PR TITLE
Enable 'Format' and 'Content' assertions during schema validation

### PR DESCRIPTION
**Pull request by JemDay**
Introduced more validation configuration options to apply during schema validation.

These changes have only currently been made to the query-parameter validation as a proof-of-concept, if this looks like an appropriate approach I will apply it more broadly - as such please don't merge art this time.

Included are some additional unit tests for query parameters to demonstrate the enforcement of 'date' and 'date-time' forma ts.

### CODE REVIEW

File: helpers/schema_compiler.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	// Decode the JSON Schema into a JSON blob.
+	decodedSchema, _ := jsonschema.UnmarshalJSON(bytes.NewReader(jsonSchema))
+	_ = compiler.AddResource(fmt.Sprintf("%s.json", name), decodedSchema)
+
+	// Try to compile it.
+	jsch, _ := compiler.Compile(fmt.Sprintf("%s.json", name))
+
+	// Done.
+	return jsch
+}
```
Comment (ccoVeille) [inline helpers/schema_compiler.go:55]: The errors are ignored here.

Is there a limitation to update the method signature to return an error?

I know the code is simply moved, but I would prefer to see it improved

Comment (JemDay) [reply helpers/schema_compiler.go:55]: Yep - I will be making this more 'error aware', I suspect the invokers of the current invokers of the function will end up ignoring the status but it will be a step in the right direction.

File: helpers/schema_compiler_test.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+		"type" : "string",
+		"enum" : [ "OTHER", "GUPPY", "PIKE", "BASS" ]
+     }
+  }
+}`
+
+func Test_SchemaWithNilOptions(t *testing.T) {
+	jsch, err := NewCompiledSchema("test", []byte(stringSchema), nil)
+
+	require.Nil(t, err, "Failed to compile Schema: %v", err)
```
Comment (ccoVeille) [inline helpers/schema_compiler_test.go:37]: Here and everywhere


```suggestion
	require.NoError(t, err, "Failed to compile Schema")
```

The asserter will report it the right way

Comment (JemDay) [reply helpers/schema_compiler_test.go:37]: I'm happy to make this change, It was a purposeful choice to use 'assert' in these situations to ensure all the issues were reported in one go as opposed to having the test exit immediately on the first failure.

File: helpers/schema_compiler_test.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+		"type" : "string",
+		"enum" : [ "OTHER", "GUPPY", "PIKE", "BASS" ]
+     }
+  }
+}`
+
+func Test_SchemaWithNilOptions(t *testing.T) {
+	jsch, err := NewCompiledSchema("test", []byte(stringSchema), nil)
+
+	require.Nil(t, err, "Failed to compile Schema")
```
Comment (ccoVeille) [inline helpers/schema_compiler_test.go:37]: You misread my previous comment

I was asking to remove err as argument, but I was asking for this


```suggestion
	require.NoError(t, err, "Failed to compile Schema")
```

Comment (JemDay) [reply helpers/schema_compiler_test.go:37]: Everyday is a learning experience ;-)

Comment (ccoVeille) [reply helpers/schema_compiler_test.go]: I suggest you enabling

testifylint linter through golangci-lint

File: config/config.go [hunk 4]
```diff
@@ -15,11 +17,17 @@ type Option func(*ValidationOptions)
...
+		ContentAssertions: false,
+	}
 
 	// Apply any supplied overrides
 	for _, opt := range opts {
-		opt(o)
+		// Sanity
+		if opt != nil {
+			opt(o)
+		}
```
Comment (ccoVeille) [inline config/config.go:30]: Maybe this


```suggestion
		if opt == nil {
 			// Sanity
			continue
		}
		opt(o)
```

The sanity is in the nil check. The comment is misleading otherwise

File: helpers/schema_compiler.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+	"github.com/pb33f/libopenapi-validator/config"
+)
+
+// ConfigureCompiler configures a JSON Schema compiler with the desired behavior.
+func ConfigureCompiler(c *jsonschema.Compiler, o *config.ValidationOptions) {
+	// Sanity
+	if o == nil {
+		return
+	}
```
Comment (ccoVeille) [inline helpers/schema_compiler.go:17]:
```suggestion
	if o == nil {
		// Sanity
		return
	}
```

File: helpers/schema_compiler_test.go [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+	// An example of a dubious JSON Schema
+	const badSchema = `{
+  "type": "you-dont-know-me",
+  "format": "date",
+  "minLength": 10
+}`
+
+	jsch, err := NewCompiledSchema("test", []byte(badSchema), nil)
+
+	assert.NotNil(t, err, "Expected an error to be thrown")
```
Comment (ccoVeille) [inline helpers/schema_compiler_test.go:76]:
```suggestion
	assert.NoError(t, err, "Expected an error to be thrown")
```

File: helpers/schema_compiler_test.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+	// An example of a JSON schema with malformed JSON
+	const badSchema = `{
+  "type": "you-dont-know-me",
+  "format": "date"
+  "minLength": 10
+}`
+
+	jsch, err := NewCompiledSchema("test", []byte(badSchema), nil)
+
+	assert.NotNil(t, err, "Expected an error to be thrown")
```
Comment (ccoVeille) [inline helpers/schema_compiler_test.go:90]:
```suggestion
	assert.NoError(t, err, "Expected an error to be thrown")
```

File: helpers/schema_compiler_test.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+     },
+     "species" : {
+      "$ref": "#/$defs/speciesEnum"
+     }
+  }
+}`
+
+	jsch, err := NewCompiledSchema("test", []byte(incompleteSchema), nil)
+
+	assert.NotNil(t, err, "Expected an error to be thrown")
```
Comment (ccoVeille) [inline helpers/schema_compiler_test.go:111]:
```suggestion
	assert.NoError(t, err, "Expected an error to be thrown")
```

File: parameters/query_parameters_test.go [hunk 9]
```diff
@@ -669,6 +670,110 @@ paths:
...
+
+	m, _ := doc.BuildV3Model()
+
+	v := NewParameterValidator(&m.Model, config.WithFormatAssertions())
+
+	request, _ := http.NewRequest(http.MethodGet, "https://things.com/a/fishy/on/a/dishy?fishy=2024-12-25", nil)
+
+	valid, errors := v.ValidateQueryParams(request)
+	assert.True(t, valid)
+	assert.Len(t, errors, 0)
```
Comment (ccoVeille) [inline parameters/query_parameters_test.go:696]:
```suggestion
	assert.Empty(t, errors)
```

File: parameters/query_parameters_test.go [hunk 10]
```diff
@@ -669,6 +670,110 @@ paths:
...
+
+	m, _ := doc.BuildV3Model()
+
+	v := NewParameterValidator(&m.Model, config.WithFormatAssertions())
+
+	request, _ := http.NewRequest(http.MethodGet, "https://things.com/a/fishy/on/a/dishy?fishy=2024-12-25T13:42:42Z", nil)
+
+	valid, errors := v.ValidateQueryParams(request)
+	assert.True(t, valid)
+	assert.Len(t, errors, 0)
```
Comment (ccoVeille) [inline parameters/query_parameters_test.go:748]:
```suggestion
	assert.Empty(t, errors)
```

File: config/config.go (hunk 4 above)
Comment (JemDay) [reply config/config.go:30]: opted (no pun intended) to remove the comment but leave the logic as-is ... seemed more logical to me.

File: helpers/schema_compiler.go (hunk 5 above)
Comment (JemDay) [reply helpers/schema_compiler.go:17]: Ack

File: helpers/schema_compiler_test.go (hunk 6 above)
Comment (JemDay) [reply helpers/schema_compiler_test.go:76]: ack

**Comment by JemDay** [comment]
> If you can bump those couple of lines, we can add this!

Will take a look.

How would I trigger those checks ... they seemed to trigger after the PR got approved

**Comment by JemDay** [comment]
Let me know how far down this rabbit hole you'd like me to go.

If I'm ready this correctly the coverage failed coverage results from some very small edge-cases around error checks (where in the past there were no checks!) - coercing tests to get the desired coverage is presumably doable but  is that a blocker ?

**Comment by daveshanley** [comment]
> Let me know how far down this rabbit hole you'd like me to go.
> 
> If I'm ready this correctly the coverage failed coverage results from some very small edge-cases around error checks (where in the past there were no checks!) - coercing tests to get the desired coverage is presumably doable but is that a blocker ?

It's not a blocker, I just hate to see coverage drop with new additions. 

I need to pull down the code and see if the work required top test it, is more work than the value its creating before waving it through. Sometimes code just needs a hall pass, other times it needs to be run through the machine.

I'll pull your branch and run the code and evaluate the level of effort to continue going down the hole or not.

**Comment by JemDay** [comment]
Thanks @daveshanley - I understand the concern and can try and go deeper as needed. I guess this is the dark-side of refactoring and adding more error-checks ;-)

**Comment by JemDay** [comment]
> Hey guys, is this still a thing? Can we hope for a merge anytime soon?

I'll try and take another look tomorrow. to see if I can push it over the line ...

**Comment by daveshanley** [comment]
I am going to pull this in, I don't want great to get in the way of good.

**Review by daveshanley — APPROVED** [review]
